invokeai/app/invocations/baseinvocation.py

```diff
@@ -3,15 +3,309 @@
+    max_digits: Optional[int] = None,
+    decimal_places: Optional[int] = None,
+    min_items: Optional[int] = None,
+    max_items: Optional[int] = None,
+    unique_items: Optional[bool] = None,
+    min_length: Optional[int] = None,
+    max_length: Optional[int] = None,
+    allow_mutation: bool = True,
+    regex: Optional[str] = None,
+    discriminator: Optional[str] = None,
+    repr: bool = True,
+    ui_type_hint: Optional[UITypeHint] = None,
+    ui_hidden: bool = False,
+    **kwargs: Any,
+) -> Any:
+    """
+    Creates an output field for an invocation output.
+
+    This is a wrapper for Pydantic's [Field](https://docs.pydantic.dev/1.10/usage/schema/#field-customization) \
+    that adds a few extra parameters to support graph execution and the node editor UI.
+
+    :param UITypeHint ui_type_hint: [None] Optionally provides an extra type hint for the UI. \
+      In some situations, the field's type is not enough to infer the correct UI type. \
+      For example, an `integer` field may be "any" integer, or a seed, which has built-in constraints. \
+      For this case, you could use `UITypeHint.Seed` to indicate that the field is a seed. \
+
+    : param bool ui_hidden: [False] Specifies whether or not this field should be hidden in the UI. \
+    """
+    return Field(
+        *args,
+        default=default,
+        default_factory=default_factory,
+        alias=alias,
+        title=title,
+        description=description,
+        exclude=exclude,
+        include=include,
+        const=const,
+        gt=gt,
+        ge=ge,
+        lt=lt,
+        le=le,
+        multiple_of=multiple_of,
+        allow_inf_nan=allow_inf_nan,
+        max_digits=max_digits,
+        decimal_places=decimal_places,
+        min_items=min_items,
+        max_items=max_items,
+        unique_items=unique_items,
+        min_length=min_length,
+        max_length=max_length,
+        allow_mutation=allow_mutation,
+        regex=regex,
+        discriminator=discriminator,
+        repr=repr,
+        ui_type_hint=ui_type_hint,
+        ui_hidden=ui_hidden,
+        **kwargs,
+    )
+
+
+InputField.__doc__ = str(InputField.__doc__) + str(Field.__doc__)
+
+
+class UIConfigBase(BaseModel):
+    """Provides additional node configuration to the UI."""
+
+    tags: Optional[list[str]] = Field(default_factory=None, description="The tags to display in the UI")
+    title: Optional[str] = Field(default=None, description="The display name of the node")
+
+
+def title(title):
+    """Adds a title to the invocation. Use this to override the default title generation, which is based on the class name."""
+
+    def wrapper(cls):
+        uiconf_name = cls.__qualname__ + ".UIConfig"
+        if not hasattr(cls, "UIConfig") or cls.UIConfig.__qualname__ != uiconf_name:
+            cls.UIConfig = type(uiconf_name, (UIConfigBase,), dict())
+        cls.UIConfig.title = title
+        return cls
+
+    return wrapper
+
+
+def tags(*tags: str):
```

Comment:
I'm not sure how to type it besides just like `Callable[..., Any]`, but that's not a particularly helpful type >.<

I tried `T = TypeVar("T", bound=BaseInvocation)` and `Callable[T, T]` but I think that's misguided (and it doesn't work)

Reply:
Yeah, it can get a little hairy and confusing pretty quickly.

I think the correct annotation here would be:
```
T = TypeVar("T", bound=BaseInvocation)

def tags(*tags: str) -> Callable[[Type[T]], Type[T]]:
    def wrapper(cls: Type[T]) -> Type[T]:
        ...
```

And then for mypy to not complain about `BaseInvocation` not having an `UIConfig` attribute, you'll need to add `UIConfig: ClassVar[Type[UIConfigBase]]` as a class level variable on `BaseInvocation`. The `ClassVar` part ensures that `pydantic` doesn't consider it as data.

Reply:
That cleared up some vagueness on my part regarding python's type annotations, and it works!

I feel all warm and fuzzy inside now, thanks :)